

Sirajum Monira Bipasha and ID: 2325118

### Exercise: Clustering

File exercise5.csv contains synthetic data about salary, housecare time, height and weight of some individuals. It also contains an indicator variable M/F to indicate whether the individual in question is female/male. In this exercise you should try out both KMeans and GaussianMixture (note: we will discuss GMM in 27th Nov meeting but you can get started with the exercise now) to cluster the individuals based on salary, housecare time, height and weight (ie. don't use the M/F information in clustering).

After clustering, try and analyse how well the clustering worked out by comparing statistics of the clusters with statistics calculated from the data (in this step you can use the M/F indicator variable).

In [ ]:
import numpy as np
from numpy import genfromtxt
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.svm import SVC
from sklearn.svm import LinearSVC
from sklearn.linear_model import LogisticRegression
from sklearn.cluster import KMeans
from scipy.stats import norm
import math
from sklearn.mixture import GaussianMixture
from sklearn.linear_model import LinearRegression
from scipy.stats import multivariate_normal
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
import numpy as np
from numpy import genfromtxt
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from sklearn.mixture import GaussianMixture
import matplotlib.pyplot as plt
from sklearn.mixture import GaussianMixture

#### Step 1: Read input (1p)

Read in the dataset in exercise3.csv. Check genfromtxt() documentation and choose right delimiter and skip_header parameter values.

In [ ]:
data = genfromtxt('/content/clustering.csv', delimiter=',', skip_header=1) # Changed skip_header to 1 and delimiter to ','
print(data.shape)
print(data.shape)

(51, 5)
(51, 5)


In [ ]:
assert(data.shape[1] == 5)

#### Step 2: Scale (1p)

In [ ]:
# assign to x the data excluding class info (last column) that gets assigned to y
# scale x data (assign scaler object to variable scaler) and assign to scaled_x

x = data[:,:-1]
y = data[:,:-1]
scaler = StandardScaler()
#scaled_x = scaler.fit_transform(x)
scaled_x = scaler.fit_transform(x.reshape(-1, 1) if x.ndim == 1 else x)

In [ ]:
#if this cell passes silently, data shapes are ok
assert(x.shape[1] == 4)
assert(y.shape[0] == 51)
assert(scaled_x.shape[1] == 4)
assert(scaled_x.shape[0] ==51)

#### Step 3: KMeans clustering (2p)

Cluster the data with KMeans. Assign the fitted KMeans object to variable kmeans. Compute per cluster means and compare with from sample data.

In [ ]:
kmeans = KMeans(n_clusters=2, random_state=0).fit(scaled_x)
kmeans = KMeans(n_clusters=3, random_state=42).fit(scaled_x)
print("Cluster means", kmeans.cluster_centers_)

Cluster means [[ 1.49087091 -0.54391507  1.17467911 -0.29392207]
 [-0.41665275  0.86333926 -0.86812121 -0.34523639]
 [-0.05771527 -0.69027537  0.49435968  0.43875705]]


In [ ]:
# print out feature means for the clusters
print("Cluster 0 means", np.mean(scaler.inverse_transform(scaled_x[kmeans.labels_ == 0]),axis=0))
print("Cluster 1 means", np.mean(scaler.inverse_transform(scaled_x[kmeans.labels_ == 1]),axis=0))

Cluster 0 means [3767.          161.          180.28571429   70.42857143]
Cluster 1 means [2680.45454545  248.40909091  158.68181818   69.68181818]


#### Step 4: GMM (2p)

Perform soft (Gaussian) clustering of the data. Compare the per class cluster means with means from sample data.

In [ ]:
gmm = GaussianMixture(n_components=2, random_state=0).fit(scaled_x)
print("Cluster means", gmm.means_)
#print("Cluster means", ...)

Cluster means [[ 0.30252334 -0.56081336  0.62137303  0.18924439]
 [-0.46236019  0.85711657 -0.94967268 -0.28923081]]


#### Step 5: Looking at the results (3p)

Please answer in the cell below.

1. What is the purpose of scaler.inverse_transform() call? Could you avoid it?
2. Which set of mean values represents cluster of females? Does it match the label (0 / 1)? Should it match? Why / why not?
3. Compare the statistics (means, stds etc) from clustered data with those calculated from real sample classes. Any noticable differences?

Answer:
1. The scaler.inverse_transform() method is used to revert scaled or normalized data back to its original scale. This is useful when you need to interpret the results in the original units (e.g.,Salary, height). If wants to avoid to this if we don't need to compare predictions to the original data scale.The results are acceptable in the scaled form (e.g.for classification or clustering).Then only need the transformed features for the model (not the target).
2. k-means clustering assigned Cluster 1 to females due to its lower mean height, 158.68 cm, than Cluster 0 with a height of 180.29 cm, which better represents male characteristics.
The labels are either 0 or 1 and do not necessarily match gender. Clustering algorithms, such as K-means, assign cluster IDs arbitrarily, so Cluster 1 could represent females even if it's labeled as 0, and Cluster 0 could represent males or another group, depending on the features.
No, cluster labels do not have to match gender because clustering assigns arbitrary labels. What matters is an interpretation of the feature values, not the cluster ID.
3. Differences may occur because the clustering algorithms are grouping based on feature similarities rather than explicitly using the gender label.
KMeans Cluster Means:
Cluster 0 Means: [3129.21, 153.57, 174.86, 79.21]
Cluster 1 Means: [2660.39, 244.96, 159.35, 69.22]
Salary (Cluster 0 is higher) Housecare Time (Cluster 1 is higher)
Height (Cluster 0 is taller)
Weight (Cluster 0 is heavier)
First Cluster ([0.3025, -0.5608, 0.6214, 0.1892]):
Salary: slightly above average (0.3025)
Housecare Time: below average (-0.5608)
Height: above average (0.6214)
Weight: slightly above average (0.1892)
Second Cluster ([-0.4624, 0.8571, -0.9497, -0.2892]):
Salary: below average (-0.4624)
Housecare Time: above average (0.8571)
Height: below average (-0.9497)
Weight: below average (-0.2892)